x/interchainquery/keeper/msg_server.go

```diff
@@ -93,6 +94,34 @@ func (k Keeper) VerifyKeyProof(ctx sdk.Context, msg *types.MsgSubmitQueryRespons
 	return nil
 }
 
+// Checks if the query timed out, and if so, checks the timeout policy
+// Returns whether the callback should be executed (only occurs if explicitly indicated)
+func (k Keeper) CheckForQueryTimeout(ctx sdk.Context, query types.Query) (executeCallback bool, err error) {
+	if !query.HasTimedOut(ctx) {
+		return true, nil
+	}
+
+	k.Logger(ctx).Error(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId,
+		"QUERY TIMEOUT - QueryId: %s, TTL: %d, BlockTime: %d", query.Id, query.TimeoutTimestamp, ctx.BlockHeader().Time.UnixNano()))
+
+	switch query.TimeoutPolicy {
+	case types.TimeoutPolicy_REJECT_QUERY_RESPONSE:
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId, "Rejecting query"))
+		return false, nil
+
+	case types.TimeoutPolicy_RETRY_QUERY_REQUEST:
+		k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(query.ChainId, query.CallbackId, "Retrying query..."))
+		return false, k.RetryICQRequest(ctx, query)
```

Comment:
If we were going to have the max retries so it doesn't infinite loop we would make it a 3rd parameter on RetryICQRequest which gets decremented each time it is called? Or were you thinking of keeping this count down on the query object itself? or alternatively make the number of retries left saved as part of the callback?
Also when the number of retries are used up, how do you specify to reject or run the callback at that point?

Reply:
yeah an extra attribute with the count on the query object is how I was thinking about it

If all retries are used up and it still times out, we should reject it.

I think if we add the retry count, it might not make sense to remove the timeout policy

Reply:
Extra attribute on the query object makes sense to me and is the cleanest option compared to the other things I was thinking like recursive calls to the function and the count param only existing in the function state.

Not only do I like the additional specification power the timeout policy gives, I think we need the timeout policy for what happens in the end when the retries run out.  Instead of 3 policy options (run_callback, fail, or retry)  We can have a policy and a number of retries.  While retries is >0 we retry but then the policy specifies what happens at the end.  And of course the number of retries can start at 0 to specify always run the callback or always fail gracefully

Reply:
On board with this solution!